debezium-testing/debezium-testing-system/src/main/java/io/debezium/testing/system/tools/databases/AbstractOcpDatabaseController.java

```diff
@@ -92,36 +89,89 @@ public String getDatabaseHostname() {
             return getDatabaseHostname();
         }
-        awaitIngress();
-
-        LoadBalancerIngress ingres = getLoadBalancedService().getStatus().getLoadBalancer().getIngress().get(0);
-        String address = (ingres.getHostname() != null) ? ingres.getHostname() : ingres.getIp();
-
-        return Objects.requireNonNull(address, "Unable to retrieve hostname or ip for service ingres");
+        return FORWARDED_HOST;
     }
 
     @Override
     public int getPublicDatabasePort() {
         if (isRunningFromOcp()) {
-            LOGGER.info("Running from OCP, using internal database port");
             return getDatabasePort();
         }
-        awaitIngress();
-        return getLoadBalancedService().getSpec().getPorts().stream()
+        return localPort;
+    }
+
+    @Override
+    public void initialize() throws InterruptedException, IOException {
+        if (!isRunningFromOcp()) {
+            forwardDatabasePorts();
+        }
+    }
+
+    @Override
+    public void forwardDatabasePorts() {
+        if (portForward != null) {
+            LOGGER.warn("Calling port forward when forward already on " + getOriginalDatabasePort() + "->" + localPort);
+            return;
+        }
+        String serviceName = getService().getMetadata().getName();
+        ServiceResource<Service> serviceResource = ocp.services().inNamespace(project).withName(serviceName);
+        int dbPort = getOriginalDatabasePort();
+        localPort = getAvailablePort();
+
+        LOGGER.info("Forwarding ports " + dbPort + "->" + localPort + " on service: " + serviceName);
+
+        PortForward forward = serviceResource
+                .portForward(dbPort, localPort);
+
+        for (Throwable e : forward.getClientThrowables()) {
+            LOGGER.error("Client error when forwarding DB port " + deployment, e);
+        }
+
+        for (Throwable e : forward.getServerThrowables()) {
+            LOGGER.error("Server error when forwarding DB port" + dbPort, e);
+        }
+        portForward = forward;
+    }
+
+    @Override
+    public void closeDatabasePortForwards() throws IOException {
+        LOGGER.info("Closing port forwards");
+        portForward.close();
+        portForward = null;
+    }
+
+    private int getOriginalDatabasePort() {
+        return getService().getSpec().getPorts().stream()
                 .filter(p -> p.getName().equals("db"))
                 .findAny()
                 .get().getPort();
     }
+
+    private int getAvailablePort() {
+        for (int i = 0; i < MAX_PORT_SEARCH_ATTEMPTS; i++) {
+            int portNum = ThreadLocalRandom.current().nextInt(MIN_PORT, MAX_PORT);
+            if (isLocalPortFree(portNum)) {
+                return portNum;
+            }
+        }
+        throw new IllegalStateException("Couldn't find free port for forwarding");
+    }
+
+    private boolean isLocalPortFree(int port) {
+        try {
+            new ServerSocket(port).close();
```

Comment:
try with resources works here.

Also note on this approach port randomization. In this particular case you are likely safe as you are using the port from the same process where this socket was previously open. In general however, this line can leave the TCP socket in state `wait_time` and make it unusable for about 2 minutes. 

Also since we are actually not using the socket but rather close it and use the formal port number, it is a subject to race possible race condition. However there is no better way unless you can pass over the socket instance directly :)

Reply:
Also -- considering this approach works here -- you can do 

```java
try( var socket = new ServerSocket(0) {
   return socket.getLocalPort();
}
```


To get a random free port assigned by the os, effectively getting rid of the random number and attempt checking generation and saving few lines of code.

Reply:
This is awesome, thanks